### PR TITLE
Move from puppeteer & looks-same to playwright & pixelmatch for more stable tests

### LINKED ISSUE #608: All visual comparing tests are failed on macbook devices
Node: 16.14.1 LTS
Platform: macOS Big Sur
Device: macbook pro late 2013

### PR DESCRIPTION
**Pull request by Kreeg**
fixes #608 as well
Tested on
* Macbook Pro M1 13" (plugged/unplugged) see here https://github.com/puppeteer/puppeteer/issues/8009
* Macbook Pro late 2013 13" (plugged/unplugged) wasn't working at all

Inspired by https://github.com/svg/svgo

TODO: compare png as buffers without additional I/O

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -45,6 +45,7 @@
     "js-yaml": "^4.1.0",
     "lodash": "^4.17.21",
     "mustache": "^4.2.0",
+    "pngjs": "^6.0.0",
```
Comment (XhmikosR) [inline package.json:48]: This is not a dependency, it should be moved to dev deps

**Comment by XhmikosR** [comment]
Also, this is really not a better solution regarding the install size. I mean sure, it installs the packages system-wide, but it still needs to download 3 different browsers + ffmpeg...

```
590 MB (619.190.067 bytes)
```

**Comment by Kreeg** [comment]
They all cropped to 1280x1024. Ok maybe it's better to deal with full-sized images, I'll check

File: test/helpers/is-pngs-matched.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+const MAX_MISMATCH = 5;
+
+/**
+ *
+ * @param {string} input                                                          input png
+ * @param {string} expected                                                       expected png
+ * @returns {Promise<{isEqual: boolean, matched: (number|*), diff: exports.PNG}>} matching results
+ */
+module.exports = async(input, expected) => {
+    const inputPng = PNG.sync.read(await fs.readFile(input));
```
Comment (XhmikosR) [inline test/helpers/is-pngs-matched.js:17]: We should use async for `PNG` too. Maybe break these 2 lines into 4?

Comment (XhmikosR) [reply test/helpers/is-pngs-matched.js:17]: NVM, I now see there's no async API.

File: package.json [hunk 3]
```diff
@@ -58,8 +58,9 @@
     "eslint-plugin-jsdoc": "^38.0.4",
     "jest": "^27.5.1",
     "less": "^4.1.2",
-    "looks-same": "^7.3.0",
```
[the remark this answers is not in the document]
Comment (Kreeg) [reply package.json:61]: * no async API
* took very long time compare, to do and save diff (more than 1 minute on old devices)

File: test/jest/setup.js [hunk 4]
```diff
@@ -1,54 +1,32 @@
...
 
 const path = require('path');
-const looksSame = require('looks-same');
-const compareSvg2PngHelper = require('../helpers/compare-svg-2-png-helper.js');
-const capturePuppeteer = require('../helpers/capture-puppeteer.js');
+const { promises: fs } = require('fs');
+const { PNG } = require('pngjs');
+const compareSvg2PngHelper = require('../helpers/compare-svg-2-png.js');
+const capturePuppeteer = require('../helpers/capture-browser.js');
+const isPngsMatched = require('../helpers/is-pngs-matched.js');
```
[the remark this answers is not in the document]
Comment (Kreeg) [reply test/jest/setup.js:8]: Hm. `compare-png-2-png` seems ok. What do you think?

[the remark this answers is not in the document]
Comment (XhmikosR) [reply test/jest/setup.js:8]: Yup, seems to match better the purpose 👍

File: package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (yisibl) [reply package.json:61]: Another advantage is that pixelmatch can diff buffer directly?

**Comment by yisibl** [comment]
I recommend using `playwright-core` or `puppeteer-core` so that you don't need to download Chrome locally, the files are very large and depending on the network status, each download is very slow.

**Comment by Kreeg** [comment]
@XhmikosR as @yisibl we can install chrome separately. But I personally don't think it's critical now (it now downloads only chromium and ffmpeg)
But if it's critical we can install only chromium in postInstall script on dev machines

**Comment by XhmikosR** [comment]
We should look into downloading any browsers only if they aren't present, but this should already be a better solution because at least playwright caches its downloads system-wide by default. Assuming it's a simple solution of course.

What I notice in this PR is that tests take twice to complete on my Windows VM.

3~0s on main vs ~60s this branch:

```
C:\Users\xmr\Desktop\svg-sprite>npm t

> svg-sprite@2.0.0-beta2 test
> jest

 PASS  test/svg-sprite/alignment/center.test.js (11.142 s)
 PASS  test/svg-sprite/minimal-configuration/modes/view.packed.test.js (11.273 s)
 PASS  test/svg-sprite/alignment/mixed.test.js (7.099 s)
 PASS  test/svg-sprite/minimal-configuration/modes/symbol.test.js (8.222 s)
 PASS  test/svg-sprite/minimal-configuration/modes/view.mixed.test.js
 PASS  test/svg-sprite/minimal-configuration/modes/css.test.js (24.117 s)
 PASS  test/svg-shape.test.js
 PASS  test/svg-sprite/minimal-configuration/no-arguments.test.js (7.852 s)
 PASS  test/spriter.test.js
 PASS  test/utils.test.js
 PASS  test/svg-sprite/minimal-configuration/error.test.js
 PASS  test/shape.test.js
 PASS  test/fix-xml-string.test.js
 PASS  test/calculate-svg-dimensions.test.js
 PASS  test/svg-sprite/minimal-configuration/rerun.test.js

Test Suites: 15 passed, 15 total
Tests:       112 passed, 112 total
Snapshots:   0 total
Time:        29.095 s, estimated 60 s
Ran all test suites.

C:\Users\xmr\Desktop\svg-sprite>git checkout -
Switched to branch 'pixelmatch'
Your branch is up to date with 'origin/pixelmatch'.

C:\Users\xmr\Desktop\svg-sprite>npm i

added 17 packages, removed 31 packages, changed 7 packages, and audited 818 packages in 3s

145 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

C:\Users\xmr\Desktop\svg-sprite>npm t

> svg-sprite@2.0.0-beta2 test
> jest

 PASS  test/svg-sprite/alignment/center.test.js (17.94 s)
 PASS  test/svg-sprite/minimal-configuration/modes/view.packed.test.js (22.875 s)
 PASS  test/svg-sprite/minimal-configuration/modes/symbol.test.js (11.752 s)
 PASS  test/svg-sprite/minimal-configuration/no-arguments.test.js (8.653 s)
 PASS  test/svg-sprite/minimal-configuration/rerun.test.js (5.762 s)
 PASS  test/svg-sprite/alignment/mixed.test.js (11.482 s)
 PASS  test/svg-shape.test.js
 PASS  test/svg-sprite/minimal-configuration/modes/view.mixed.test.js (6.664 s)
 PASS  test/spriter.test.js
 PASS  test/shape.test.js
 PASS  test/svg-sprite/minimal-configuration/error.test.js
 PASS  test/calculate-svg-dimensions.test.js
 PASS  test/fix-xml-string.test.js
 PASS  test/utils.test.js
 PASS  test/svg-sprite/minimal-configuration/modes/css.test.js (57.867 s)

Test Suites: 15 passed, 15 total
Tests:       112 passed, 112 total
Snapshots:   0 total
Time:        58.456 s
Ran all test suites.
```

File: test/svg-sprite/alignment/center.test.js [hunk 5]
```diff
@@ -177,7 +177,7 @@ describe(`svg-sprite: with centered alignment and ${align.length} SVG files`, ()
             expect.hasAssertions();
 
             const input = path.join(tmpPath, 'css/svg', svgPath);
-            const expected = path.join(paths.expectations, 'png/css.packed.aligned.png');
+            const expected = path.join(paths.expectations, 'png/css.packed.centered.png');
```
Comment (XhmikosR) [inline test/svg-sprite/alignment/center.test.js:180]: Maybe you could/should split this change and the view.mixed.test.js change to a separate PR?

**Comment by Kreeg** [comment]
Hm... that's strange. Here is my result for macbook pro m1

```
 ~/projects/svg-sprite   main  npm run test

> svg-sprite@2.0.0-beta2 test
> jest

 PASS  test/svg-sprite/minimal-configuration/modes/view.mixed.test.js (11.329 s)
 PASS  test/svg-sprite/minimal-configuration/no-arguments.test.js (14.677 s)
 PASS  test/svg-shape.test.js
 PASS  test/svg-sprite/alignment/mixed.test.js (16.294 s)
 PASS  test/svg-sprite/minimal-configuration/modes/symbol.test.js (16.693 s)
 PASS  test/svg-sprite/minimal-configuration/error.test.js
 PASS  test/shape.test.js
 PASS  test/spriter.test.js
 PASS  test/calculate-svg-dimensions.test.js
 PASS  test/fix-xml-string.test.js
 PASS  test/svg-sprite/minimal-configuration/rerun.test.js (7.022 s)
 PASS  test/utils.test.js
 PASS  test/svg-sprite/minimal-configuration/modes/view.packed.test.js (20.411 s)
 PASS  test/svg-sprite/alignment/center.test.js (20.395 s)
 PASS  test/svg-sprite/minimal-configuration/modes/css.test.js (36.68 s)

Test Suites: 15 passed, 15 total
Tests:       112 passed, 112 total
Snapshots:   0 total
Time:        37.904 s
Ran all test suites.



 ~/projects/svg-sprite   main  git switch pixelmatch
Switched to branch 'pixelmatch'
Your branch is up to date with 'origin/pixelmatch'.
 ~/projects/svg-sprite   pixelmatch  npm i

added 17 packages, removed 31 packages, changed 7 packages, and audited 819 packages in 3s

145 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
 ~/projects/svg-sprite   pixelmatch  npm run test

> svg-sprite@2.0.0-beta2 test
> jest

 PASS  test/svg-sprite/minimal-configuration/modes/view.mixed.test.js (7.473 s)
 PASS  test/svg-sprite/minimal-configuration/no-arguments.test.js (8.326 s)
 PASS  test/spriter.test.js
 PASS  test/calculate-svg-dimensions.test.js
 PASS  test/svg-sprite/minimal-configuration/modes/symbol.test.js (11.371 s)
 PASS  test/svg-sprite/alignment/mixed.test.js (12.372 s)
 PASS  test/svg-shape.test.js
 PASS  test/shape.test.js
 PASS  test/svg-sprite/minimal-configuration/rerun.test.js (5.463 s)
 PASS  test/utils.test.js
 PASS  test/fix-xml-string.test.js
 PASS  test/svg-sprite/minimal-configuration/error.test.js
 PASS  test/svg-sprite/alignment/center.test.js (15.813 s)
 PASS  test/svg-sprite/minimal-configuration/modes/view.packed.test.js (15.891 s)
 PASS  test/svg-sprite/minimal-configuration/modes/css.test.js (32.881 s)

Test Suites: 15 passed, 15 total
Tests:       112 passed, 112 total
Snapshots:   0 total
Time:        33.655 s, estimated 37 s
Ran all test suites.
```
I'll check it now on mackbook pro 2013

**Comment by Kreeg** [comment]
I added feature for reusing chromium instance, this accelerated tests a bit

**Comment by XhmikosR** [comment]
True, still slower on Windows compared to main :/

```
C:\Users\xmr\Desktop\svg-sprite>npm t

> svg-sprite@2.0.0-beta2 test
> jest

 PASS  test/svg-sprite/alignment/center.test.js (11.987 s)
 PASS  test/svg-sprite/minimal-configuration/modes/view.packed.test.js (16.681 s)
 PASS  test/svg-sprite/minimal-configuration/modes/symbol.test.js (9.831 s)
 PASS  test/svg-sprite/alignment/mixed.test.js (8.679 s)
 PASS  test/svg-sprite/minimal-configuration/modes/view.mixed.test.js (5.061 s)
 PASS  test/svg-sprite/minimal-configuration/no-arguments.test.js (8.633 s)
 PASS  test/utils.test.js
 PASS  test/svg-shape.test.js
 PASS  test/calculate-svg-dimensions.test.js
 PASS  test/svg-sprite/minimal-configuration/rerun.test.js (6.316 s)
 PASS  test/svg-sprite/minimal-configuration/error.test.js
 PASS  test/shape.test.js
 PASS  test/fix-xml-string.test.js
 PASS  test/spriter.test.js
 PASS  test/svg-sprite/minimal-configuration/modes/css.test.js (45.504 s)

Test Suites: 15 passed, 15 total
Tests:       112 passed, 112 total
Snapshots:   0 total
Time:        45.939 s, estimated 48 s
Ran all test suites.
```

Not sure if there's something we can do, I still don't get the root cause compared to main...

**Comment by Kreeg** [comment]
The results for macbook pro late 2013 are:
```
Current branch pixelmatch is up to date.
 ~/projects/svg-sprite   pixelmatch  npm run test

> svg-sprite@2.0.0-beta2 test
> jest

 PASS  test/svg-sprite/alignment/center.test.js (57.068 s)
 PASS  test/svg-sprite/minimal-configuration/modes/view.packed.test.js (72.375 s)
 PASS  test/svg-sprite/alignment/mixed.test.js (39.709 s)
 PASS  test/svg-sprite/minimal-configuration/modes/symbol.test.js (38.138 s)
 PASS  test/svg-sprite/minimal-configuration/no-arguments.test.js (26.603 s)
 PASS  test/svg-sprite/minimal-configuration/modes/view.mixed.test.js (17.498 s)
 PASS  test/svg-shape.test.js
 PASS  test/svg-sprite/minimal-configuration/error.test.js
 PASS  test/svg-sprite/minimal-configuration/rerun.test.js (14.77 s)
 PASS  test/spriter.test.js
 PASS  test/shape.test.js
 PASS  test/calculate-svg-dimensions.test.js
 PASS  test/fix-xml-string.test.js
 PASS  test/utils.test.js
 PASS  test/svg-sprite/minimal-configuration/modes/css.test.js (163.657 s)

Test Suites: 15 passed, 15 total
Tests:       112 passed, 112 total
Snapshots:   0 total
Time:        164.6 s, estimated 201 s
Ran all test suites.

Cant check passing tests on main branch cause they are all failing
Test Suites: 6 failed, 9 passed, 15 total
Tests:       28 failed, 84 passed, 112 total
Snapshots:   0 total
Time:        869.47 s
Ran all test suites.
```

At least the playwright version is more stable. Maybe this is one case of many problems with puppeteer (it has 1.8k opened issues against 460+ in playwright)

**Comment by XhmikosR** [comment]
I don't suppose this has to do with moving the code outside of jest.setup.js?

It's not a huge deal, but it'd be nice if we didn't introduce such slowdowns.

I'll merge this since it fixes your issue, but see if we can find any way to speed this up again later (lower priority compared to our issues ofc :))

**Review by XhmikosR — APPROVED** [review]

**Comment by XhmikosR** [comment]
@Kreeg OK, I think it has to do with the fact that we are now spawning too many processes, which is known to be slow on Windows.

With this patch I see spikes to the number of processes up to ~280, while before this patch, this was maxed at ~120 processes.

**Comment by Kreeg** [comment]
@XhmikosR how do you measure it? Can you check with `npx jest --maxWorkers=50%`?

**Comment by XhmikosR** [comment]
I see the number of processes during tests, there's no magic :)

Using `maxWorkers` won't fix the issue and it's be slower, it's the logic that we are running something asynchronously here which in turn spawns too many processes.

```
C:\Users\xmr\Desktop\svg-sprite>npm t -- --maxWorkers=50%

> svg-sprite@2.0.0-beta2 test
> jest "--maxWorkers=50%"

 PASS  test/svg-sprite/minimal-configuration/modes/view.packed.test.js (14.427 s)
 PASS  test/svg-sprite/alignment/center.test.js (9.231 s)
 PASS  test/svg-sprite/minimal-configuration/modes/symbol.test.js (8.84 s)
 PASS  test/svg-sprite/minimal-configuration/modes/css.test.js (39.915 s)
 PASS  test/svg-sprite/alignment/mixed.test.js (7.612 s)
 PASS  test/svg-sprite/minimal-configuration/modes/view.mixed.test.js
 PASS  test/svg-sprite/minimal-configuration/no-arguments.test.js (7.792 s)
 PASS  test/svg-shape.test.js
 PASS  test/svg-sprite/minimal-configuration/rerun.test.js (5.412 s)
 PASS  test/shape.test.js
 PASS  test/spriter.test.js
 PASS  test/svg-sprite/minimal-configuration/error.test.js
 PASS  test/calculate-svg-dimensions.test.js
 PASS  test/utils.test.js
 PASS  test/fix-xml-string.test.js

Test Suites: 15 passed, 15 total
Tests:       112 passed, 112 total
Snapshots:   0 total
Time:        56.486 s, estimated 63 s
Ran all test suites.
```

**Comment by Kreeg** [comment]
Yes, this code has more async/await than before. I thought this is our ultima thule :)

**Comment by XhmikosR** [comment]
It is, but it seems we overdid with with browsers :/

Everywhere else it should be fine, but here it has a negative impact. Not a huge deal, but it'd be nice if we could bring this back back the previous times for tests :)

**Comment by XhmikosR** [comment]
Actually, the slowness is noticeable on Ubuntu on CI too. Unsure if it's only due to async or it's playwright.

If you could figure this out, that's be perfect :)

**Pull request merged**